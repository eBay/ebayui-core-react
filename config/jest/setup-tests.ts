import '@testing-library/jest-dom'

declare var global: typeof globalThis;

if (typeof window.URL.createObjectURL === 'undefined') {
  Object.defineProperty(window.URL, 'createObjectURL', { value: () => {} });
}
