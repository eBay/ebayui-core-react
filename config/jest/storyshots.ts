import { render } from "@testing-library/react";
import initStoryshotsAddon from "@storybook/addon-storyshots";
import type { StoryshotsOptions } from "@storybook/addon-storyshots/dist/ts3.9/api/StoryshotsOptions";

const reactTestingLibrarySerializer = {
    print: (val, serialize, indent) => serialize(val.asFragment()),
    test: (val) => val && val.hasOwnProperty("container"),
};

export function initStoryshots(config: StoryshotsOptions) {
    initStoryshotsAddon({
        ...config,
        renderer: render,
        snapshotSerializers: [reactTestingLibrarySerializer],
    });
}
