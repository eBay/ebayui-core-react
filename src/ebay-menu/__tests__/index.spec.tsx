import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { initStoryshots } from '../../../config/jest/storyshots'
import { EbayMenu, EbayMenuItem } from '../index'

initStoryshots({
    config: ({ configure }) => configure(() => require('./index.stories'), module)
})

describe('<EbayMenu>', () => {
    describe('on menu item click', () => {
        it('should fire onClick event', () => {
            const onClickSpy = jest.fn()
            const wrapper = render(
                <EbayMenu>
                    <EbayMenuItem onClick={onClickSpy}/>
                </EbayMenu>
            )

            fireEvent.click(wrapper.container.querySelector('.menu__item'))

            expect(onClickSpy).toBeCalled()
        })
    })
    describe('on key down', () => {
        it('should fire onKeyDown event', () => {
            const onKeyDownSpy = jest.fn()
            const wrapper = render(
                <EbayMenu onKeyDown={onKeyDownSpy}>
                    <EbayMenuItem/>
                </EbayMenu>
            )
            fireEvent.keyDown(wrapper.container.querySelector('.menu__item'))

            expect(onKeyDownSpy).toBeCalledWith(0, false)
        })
    })
    describe('on radio item select', () => {
        it('should fire onSelect & onChange event', () => {
            const onSelectSpy = jest.fn()
            const onChangeSpy = jest.fn()
            const wrapper = render(
                <EbayMenu onSelect={onSelectSpy} onChange={onChangeSpy} type='radio'>
                    <EbayMenuItem checked />
                    <EbayMenuItem />
                </EbayMenu>
            )
            expect(wrapper.container.querySelector('.menu__item')).toHaveAttribute('aria-checked', 'true')
            expect(wrapper.container.querySelectorAll('.menu__item')[1]).toHaveAttribute('aria-checked', 'false')

            fireEvent.click(wrapper.container.querySelectorAll('.menu__item')[1])

            expect(onChangeSpy).toBeCalledWith(1, true)
            expect(onSelectSpy).toBeCalledWith(1, true)

            expect(wrapper.container.querySelector('.menu__item')).toHaveAttribute('aria-checked', 'false')
            expect(wrapper.container.querySelectorAll('.menu__item')[1]).toHaveAttribute('aria-checked', 'true')
        })
    })
    describe('on checkbox item select', () => {
        it('should fire onSelect & onChange event', () => {
            const onSelectSpy = jest.fn()
            const onChangeSpy = jest.fn()
            const wrapper = render(
                <EbayMenu onSelect={onSelectSpy} onChange={onChangeSpy} type='checkbox'>
                    <EbayMenuItem />
                    <EbayMenuItem checked />
                </EbayMenu>
            )
            expect(wrapper.container.querySelector('.menu__item')).not.toHaveAttribute('aria-checked')
            expect(wrapper.container.querySelectorAll('.menu__item')[1]).toHaveAttribute('aria-checked', 'true')

            fireEvent.click(wrapper.container.querySelectorAll('.menu__item')[1])

            expect(onChangeSpy).toBeCalledWith(1, false)
            expect(onSelectSpy).toBeCalledWith(1, false)

            expect(wrapper.container.querySelector('.menu__item')).not.toHaveAttribute('aria-checked')
            expect(wrapper.container.querySelectorAll('.menu__item')[1]).toHaveAttribute('aria-checked', 'false')

            fireEvent.click(wrapper.container.querySelectorAll('.menu__item')[0])

            expect(wrapper.container.querySelector('.menu__item')).toHaveAttribute('aria-checked', 'true')
            expect(wrapper.container.querySelectorAll('.menu__item')[1]).toHaveAttribute('aria-checked', 'false')
        })
    })
})
