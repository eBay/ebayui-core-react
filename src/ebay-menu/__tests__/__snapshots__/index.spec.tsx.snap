// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots ebay-menu Checkbox 1`] = `
<DocumentFragment>
  <span
    class="menu"
  >
    <div
      class="menu__items"
      role="menu"
    >
      <div
        aria-checked="true"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
        value="item 1"
      >
        <span>
          item 1
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
      <div
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
        value="item 2"
      >
        <span>
          item 2
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
      <div
        aria-checked="true"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
        value="item 3"
      >
        <span>
          item 3
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-menu Default 1`] = `
<DocumentFragment>
  <span
    class="menu"
  >
    <div
      class="menu__items"
      role="menu"
    >
      <div
        aria-checked="false"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
      >
        <span>
          Item 1 that has very long text
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
      <div
        aria-checked="false"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
      >
        <span>
          Item 2
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
      <div
        aria-checked="false"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
      >
        <span>
          Item 3
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-menu Radio 1`] = `
<DocumentFragment>
  <span
    class="menu"
  >
    <div
      class="menu__items"
      role="menu"
    >
      <div
        aria-checked="true"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
        value="item 1"
      >
        <span>
          item 1
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
      <div
        aria-checked="false"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
        value="item 2"
      >
        <span>
          item 2
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
      <div
        aria-checked="false"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
        value="item 3"
      >
        <span>
          item 3
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-menu With Badges 1`] = `
<DocumentFragment>
  <span
    class="menu"
  >
    <div
      class="menu__items"
      role="menu"
    >
      <div
        aria-checked="false"
        aria-hidden="true"
        aria-label="item 1 (5 unread items)"
        class="menu__item menu__item--badged"
        role="menuitem"
        tabindex="0"
      >
        <span>
          item 1
          <span
            aria-hidden="true"
            class="badge"
          >
            5
          </span>
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
      <div
        aria-checked="false"
        aria-hidden="true"
        aria-label="item 2 (23 unread items)"
        class="menu__item menu__item--badged"
        role="menuitem"
        tabindex="0"
      >
        <span>
          item 2
          <span
            aria-hidden="true"
            class="badge"
          >
            23
          </span>
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
      <div
        aria-checked="false"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
      >
        <span>
          item 3
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-menu With Disabled Item 1`] = `
<DocumentFragment>
  <span
    class="menu"
  >
    <div
      class="menu__items"
      role="menu"
    >
      <div
        aria-checked="false"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
      >
        <span>
          item 1 that has very long text
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
      <div
        aria-checked="false"
        aria-disabled="true"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
      >
        <span>
          Item 2
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
      <div
        aria-checked="false"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
      >
        <span>
          Item 3
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-menu With Icons 1`] = `
<DocumentFragment>
  <span
    class="menu"
  >
    <div
      class="menu__items"
      role="menu"
    >
      <div
        aria-checked="false"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
      >
        <span>
          <svg
            aria-hidden="true"
            class="icon icon--confirmation"
            focusable="false"
            height="24px"
            style="margin-right: 8px;"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="#icon-confirmation"
            />
          </svg>
           Confirmed
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
      <div
        aria-checked="false"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
        value="item 2"
      >
        <span>
          <svg
            aria-hidden="true"
            class="icon icon--attention"
            focusable="false"
            height="24px"
            style="margin-right: 8px;"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="#icon-attention"
            />
          </svg>
           Not yet confirmed
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
      <div
        aria-checked="false"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
        value="item 3"
      >
        <span>
          <svg
            aria-hidden="true"
            class="icon icon--attention"
            focusable="false"
            height="24px"
            style="margin-right: 8px;"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="#icon-attention"
            />
          </svg>
           Not yet confirmed
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-menu With Separator 1`] = `
<DocumentFragment>
  <span
    class="menu"
  >
    <div
      class="menu__items"
      role="menu"
    >
      <div
        aria-checked="false"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
      >
        <span>
          item 1 that has very long text
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
      <div
        aria-checked="false"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
      >
        <span>
          Item 2
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
      <hr
        class="menu__separator"
        role="separator"
        tabindex="0"
      />
      <div
        aria-checked="false"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
      >
        <span>
          Item 3
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
      <div
        aria-checked="false"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
      >
        <span>
          Item 4
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
      <div
        aria-checked="false"
        aria-hidden="false"
        class="menu__item"
        role="menuitem"
        tabindex="0"
      >
        <span>
          Item 5
        </span>
        <svg
          aria-hidden="true"
          class="icon icon--tick-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-tick-small"
          />
        </svg>
      </div>
    </div>
  </span>
</DocumentFragment>
`;
